slice_test.go

```diff
@@ -919,3 +923,394 @@ func TestMetrics(t *testing.T) {
+		[]string{"zero", "one", "two", "three"},
+	)
+	slicetest.Print(slice)
+	// Output:
+	// 0 zero
+	// 1 one
+	// 2 two
+	// 3 three
+}
+
+func ExampleFilter() {
+	slice := bigslice.Const(2,
+		[]int{0, 1, 2, 3, 4, 5},
+		[]string{"zero", "one", "two", "three", "four", "five"},
+	)
+	slice = bigslice.Filter(slice, func(x int, s string) bool {
+		return x%2 == 0
+	})
+	slicetest.Print(slice)
+	// Output:
+	// 0 zero
+	// 2 two
+	// 4 four
+}
+
+func ExampleFlatmap() {
+	slice := bigslice.Const(2,
+		[]string{
+			"Lorem ipsum dolor sit amet",
+			"consectetur:adipiscing",
+			"elit",
+			"sed.do.eiusmod.tempor.incididunt",
+		},
+		[]string{" ", ":", ";", "."}, // Separators.
+	)
+	slice = bigslice.Flatmap(slice, func(s, sep string) ([]string, []int) {
+		split := strings.Split(s, sep)
+		lengths := make([]int, len(split))
+		for i := range lengths {
+			lengths[i] = len(split[i])
+		}
+		return split, lengths
+	})
+	slicetest.Print(slice)
+	// Output:
+	// Lorem 5
+	// adipiscing 10
+	// amet 4
+	// consectetur 11
+	// do 2
+	// dolor 5
+	// eiusmod 7
+	// elit 4
+	// incididunt 10
+	// ipsum 5
+	// sed 3
+	// sit 3
+	// tempor 6
+}
+
+func ExampleFold() {
+	slice := bigslice.Const(2,
+		[]string{"c", "a", "b", "c", "c", "b", "a", "a", "a", "a", "c"},
+		[]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11},
+		[]string{
+			"Lorem",
+			"ipsum",
+			"dolor",
+			"sit",
+			"amet",
+			"consectetur",
+			"adipiscing",
+			"elit",
+			"sed",
+			"do",
+			"eiusmod",
+		},
+	)
+	type accum struct {
+		ready bool
+		// sum is the sum of integers in the second column.
+		sum int
+		// product is the prouct of integers in the second column.
+		prod int
+		// longest is the longest string in the third column.
+		longest string
+	}
+	slice = bigslice.Fold(slice, func(acc accum, i int, s string) accum {
+		if !acc.ready {
+			acc.prod = 1
+			acc.ready = true
+		}
+		acc.sum += i
+		acc.prod *= i
+		if len(acc.longest) < len(s) {
+			acc.longest = s
+		}
+		return acc
+	})
+	slicetest.Print(slice)
+	// Output:
+	// a {true 36 10080 adipiscing}
+	// b {true 9 18 consectetur}
+	// c {true 21 220 eiusmod}
+}
+
+func ExampleHead() {
+	// Use one shard, as Head operates per shard.
+	slice := bigslice.Const(1,
+		[]int{0, 1, 2, 3, 4, 5},
+		[]string{"zero", "one", "two", "three", "four", "five"},
+	)
+	slice = bigslice.Head(slice, 3)
+	slicetest.Print(slice)
+	// Output:
+	// 0 zero
+	// 1 one
+	// 2 two
+}
+
+func ExampleMap() {
+	slice := bigslice.Const(2,
+		[]int{0, 1, 2, 3},
+		[]string{"zero", "one", "two", "three"},
+	)
+	slice = bigslice.Map(slice, func(x int, s string) (int, int, string) {
+		return x, x * x, fmt.Sprintf("%s.squared", s)
+	})
+	slicetest.Print(slice)
+	// Output:
+	// 0 0 zero.squared
+	// 1 1 one.squared
+	// 2 4 two.squared
+	// 3 9 three.squared
+}
+
+func ExamplePrefixed() {
+	slice := bigslice.Const(2,
+		[]string{"a", "a", "a", "a", "a", "b", "b", "b", "b", "c", "c"},
+		[]string{"c", "a", "b", "c", "c", "b", "a", "a", "a", "a", "c"},
```

Comment:
Optional: Use different letters here, or maybe two categories of objects (cat, dog, etc. vs. red, green, blue) for the two slice columns so the output is a little less alphabet-soupy.

Reply:
I switched it to count the number of pet (type, name) pairs.